src/game/game_state.cpp

```diff
@@ -1,67 +1,77 @@
 #include "game_state.h"
 #include "events.h"
 
-MenuPhase GameState::menu_phase;
-BuildPhase GameState::build_phase;
-DungeonPhase GameState::dungeon_phase;
-MinigamePhase GameState::minigame_phase;
+Context GameState::context;
+MenuPhase GameState::menu_phase(context);
+BuildPhase GameState::build_phase(context);
+DungeonPhase GameState::dungeon_phase(context);
+MinigamePhase GameState::minigame_phase(context);
 Phase* GameState::curr_phase;
 std::map<int, Player*> GameState::players;
 
 Physics GameState::physics;
 SceneManager GameState::scene_manager;
 MainScene GameState::testScene;
 int GameState::num_players = 0;
+bool GameState::is_server = true;
 
-void GameState::init(Phase* phase) {
-    curr_phase = phase;
-    curr_phase->setup();
-
+void GameState::init() {
     physics.set_scene(&testScene);
     scene_manager.set_current_scene(&testScene);
 
+    // Client of given connection id wishes to join the game.
     events::menu::request_join_event.connect([](int conn_id) {
         proto::JoinResponse resp;
         resp.set_status(num_players < 4);
         resp.set_id(conn_id);
-        events::menu::respond_join_event(conn_id, resp);
 
-        if (num_players < 4)
-            add_player(conn_id, false);
+        // If the game isn't full (< 4 players), create a Player object for the client.
+        if (num_players < 4) {
+            Player* player = add_new_player();
+            resp.set_obj_id(player->get_id());
+            players[conn_id] = player;
+            num_players++;
+        }
+        events::menu::respond_join_event(conn_id, resp);
     });
 
-    events::menu::spawn_player_event.connect([](proto::Player & p) {
-        add_player(p.id(), true);
+    events::menu::spawn_existing_player_event.connect([](int id) {
+        add_existing_player(id);
     });
 }
 
 void GameState::graphical_setup() {
+    is_server = false;
```

Comment:
Maybe the determinant of whether it's a server or a client shouldn't hinge upon its calling graphical_setup.